### DISCUSSION TITLE
Get Data via MQTT

### DISCUSSION THREAD

**Discussion by Werner1000** [meatpiHQ/wican-fw#190] [Q&A]
WICAN works fine with Car Scanner App via BT and WIFI, so I was able to collect the SoC PID for my VW E-UP via the logging function on MQTT:
wican/xxxxxxxxxxxx/elm327 {"bus":"0","type":"tx","ts":35144,"frame":[{"id":2016,"dlc":8,"rtr":false,"extd":false,"data":[3,34,22,24,170,170,170,170]}]}
wican/xxxxxxxxxxxx/elm327 {"bus":"0","type":"rx","ts":35146,"frame":[{"id":2024,"dlc":8,"rtr":false,"extd":false,"data":[5,98,22,24,37,88,85,85]}]}

To request the SoC via MQTT I sent this request:
mosquitto_pub -h 192.168.xx.x -p 1883 -u wican -P wican -t wican/xxxxxxxxxxxx/can/tx -m "{"bus":"0","type":"tx","frame":[{"id":2016,"dlc":8,"rtr":false,"extd":false,"data":[3,34,22,24,170,170,170,170]}]}"

There is absolutely no reponse from the WICAN, even if I try several proposals / different PIDs available from different online-sources. My WICAN is functional connected to my MQTT Broker, the ONLINE- and OFFLINE messages are published as the logging-functions. WICAN SW-Version is the actual 3.28.

Any ideas whats wrong with my approach? Any hints are welcome!
Best regards
Werner

**Comment by cr0i** [comment]
> WICAN works fine with Car Scanner App via BT and WIFI, so I was able to collect the SoC PID for my VW E-UP via the logging function on MQTT: wican/xxxxxxxxxxxx/elm327 {"bus":"0","type":"tx","ts":35144,"frame":[{"id":2016,"dlc":8,"rtr":false,"extd":false,"data":[3,34,22,24,170,170,170,170]}]} wican/xxxxxxxxxxxx/elm327 {"bus":"0","type":"rx","ts":35146,"frame":[{"id":2024,"dlc":8,"rtr":false,"extd":false,"data":[5,98,22,24,37,88,85,85]}]}
> 
> To request the SoC via MQTT I sent this request: mosquitto_pub -h 192.168.xx.x -p 1883 -u wican -P wican -t wican/xxxxxxxxxxxx/can/tx -m "{"bus":"0","type":"tx","frame":[{"id":2016,"dlc":8,"rtr":false,"extd":false,"data":[3,34,22,24,170,170,170,170]}]}"
> 
> There is absolutely no reponse from the WICAN, even if I try several proposals / different PIDs available from different online-sources. My WICAN is functional connected to my MQTT Broker, the ONLINE- and OFFLINE messages are published as the logging-functions. WICAN SW-Version is the actual 3.28.
> 
> Any ideas whats wrong with my approach? Any hints are welcome! Best regards Werner

Elm Logging must be disabled again. 
Perhaps you can try with ignition on? 
That was a problem as I tried the first times, but without HA.

**Comment by meatpiHQ** [comment]
@Werner1000 I've just created a vehicle profile for e-up, please update to the latest firmware and try it

https://github.com/meatpiHQ/wican-fw/releases
https://github.com/meatpiHQ/wican-fw?tab=readme-ov-file#getting-started-guide

**Reply by Werner1000** [reply]
Just updated to 3.30 plus new profile, made automation settings and - magic - two PIDs are automatically published via MQTT:
wican/xxxxxxxxxx/can/rx {"bus":"0","type":"rx","ts":203,"frame":[{"id":2029,"dlc":8,"rtr":false,"extd":false,"data":[6,65,0,152,24,0,1,170]}]}
wican/xxxxxxxxxx/can/rx {"bus":"0","type":"rx","ts":206,"frame":[{"id":2024,"dlc":8,"rtr":false,"extd":false,"data":[6,65,0,128,8,0,17,0]}]}
Exactly what I'm looking for.

PID 2029 is continueously published, even after switching off the ignition, PID 2024 only when ignition is switched on.

Was not shure about automation settings, with these ones it worked well for me:
![Automation Settings](https://github.com/user-attachments/assets/3b3a2600-9cfe-41d0-aad9-24d728ad566c)

Based on this I will continue to implement it to my environment.
Thanks guys!

**Reply by meatpiHQ** [reply]
@Werner1000 you should receive the SoC value in this format  {"SOC": "xx"}

The profile is only sending one parameter, where is the other one coming from? Do you have a custom PID?

**Comment by Werner1000** [comment]
Sad to say, seems to have been a dayfly yesterday. meatpi was switched off overnight, today same procedure as before with ver 2.98.: Only 2 MQTT-messages came when reboot/start ('MQTT elm327 log' disabled) - nothing else: 
wican/xxxxxxxxxxxx/can/status {"status": "offline"}
wican/xxxxxxxxxxxx/can/status {"status": "online"}

and if MQTT elm327 log is enabled and Car Scanner is active:
wican/xxxxxxxxxxx/elm327 {"bus":"0","type":"rx","ts":20961,"frame":[{"id":2029,"dlc":8,"rtr":false,"extd":false,"data":[6,65,64,224,128,0,0,170]}]}
wican/xxxxxxxxxxx/elm327 {"bus":"0","type":"rx","ts":20965,"frame":[{"id":2024,"dlc":8,"rtr":false,"extd":false,"data":[6,65,64,196,195,128,33,0]}]}
wican/xxxxxxxxxxx/elm327 {"bus":"0","type":"tx","ts":21596,"frame":[{"id":2015,"dlc":8,"rtr":false,"extd":false,"data":[2,1,96,170,170,170,170,170]}]}

 2 questions to the 'Automate' settings: 
- is 'Destination Topic:' in automate settings the full MQTT topic or only an addition to the general MQTT-settings?
- for pure MQTT-publishing without HA usage is 'Home Assistant Discovery:' disabled sufficiant?

Do I have any possibilty to get more detailed logs ?

Best regards
Werner

**Reply by cr0i** [reply]
With Automate Settings it seems to be necessary that the Initialisation string sets the Header (ID) for the Request/Response.
Because in the moment only one Initialisation string is possible I only can get one value published.
I was wondering, why you could store the settings without Destination Topic, it was mandatory at me.
When you use Automate (With AutoPID Protokoll), then you don´t send the tx frame from your system, the wican publishes the MQTT Topic itself.

Based on my Init String, try changing yours in the car Profile to something like that:
ATSP6;ATST96;ATSH7E5;ATCP18;ATCRA7ED;ATFCSH7E5;ATFCSD300000;ATFCSM1;
(I´m not sure if all are compatible at you, but I think important would be: ATSH7E5;ATCRA7ED;ATFCSH7E5;)
Then give an Destination Topic, under this topic should a new topic SOC be published if Protokoll changed to AutoPID and the wican is rebooted.
So I got it working.

**Reply by meatpiHQ** [reply]
@Werner1000 what firmware version are you running?

**Reply by Werner1000** [reply]
Again thanks to you both, now it looks good. I get these MQTT-Messages at v3.30 (hopefully permanently...)

wican/xxxxxxxxxx/can/rx {"bus":"0","type":"rx","ts":40414,"frame":[{"id":2029,"dlc":8,"rtr":false,"extd":false,"data":[4,98,2,140,233,170,170,170]}]}
wican/xxxxxxxxxx/can/SoC {"SOC":93.2}

Only these two adaptions have been made at my side, no need for the proposals from cr0i. Everything works as default.
- Change the Settings/Protocol from elm327 to Autopid (my error, read the f... manual). Consequence: Car Scanner needs elm327, has to be changed when in use.
- add the full MQTT topic at Automate/destination Topic: wican/xxxxxxxxx/can/SoC, otherwise it will publish to complete different topic out of my focus. Proposal for future versions: why not use default values similar to these at Settings/MQTT?
- and of cause: disable Settings/elm327 logging

**Reply by meatpiHQ** [reply]
@Werner1000 alright so we can say that E-Up profile works.

> Proposal for future versions: why not use default values similar to these at Settings/MQTT?

if you leave destination empty, where does it publish? It's suppose to send to the default topic if left empty.

**Reply by Werner1000** [reply]
Yes, for me E-Up profile now works. I emptied Automate/Destination path and now it publishes as you desripted along settings/MQTT . wican/xxxxxxxxxxxx/can/SoC {"SOC":95.6}
Thats fine, maybe note it somewhere visible - it was not clear to me what's needed.

Remark: calculated SoC is too low: Real 100% SoC is published as 95,6%. Not a problem for me, I multiply the value with 1,046 and thats sufficiant for me.

**Reply by schuepc** [reply]
We had a long discussion about this topic in the evcc-io project (see: [Discussion #18744 – OBD integration Škoda Citigo-e iV](https://github.com/evcc-io/evcc/discussions/18744).
The original thread was in German – here is the English translation and my current working setup:

The code below is running. However, I still need to apply a +2% offset to the SOC so that it matches the value shown in the Škoda App at 80%. I will continue monitoring this in the next few days.

I also managed to read out the range now – you just have to wait a little until the value comes in.

Regarding the codes, this resource helped me a lot: [Liste der OBD2-Codes (GoingElectric Wiki)](https://www.goingelectric.de/wiki/Liste-der-OBD2-Codes)

FW Version tested: [wican-fw_obd_v412-102-ge1d0a09.zip](https://github.com/user-attachments/files/21836606/wican-fw_obd_v412-102-ge1d0a09.zip)

My working configuration
```
{
  "cars": [
    {
      "car_model": "Škoda Citigo-e iV",
      "init": "ATSP6;ATST96;ATSH7ED;ATSH714;",
      "pids": [
        {
          "pid": "22 02 8C",
          "pid_init": "ATSH7E5;",
          "parameters": [
            { "name": "SOC", "expression": "B4/2.55+2", "unit": "%", "class": "battery", "min": "0", "max": "100", "period": 5000 }
          ]
        },
        {
          "pid": "22 02 BD",
          "pid_init": "ATSH7E5;",
          "parameters": [
            { "name": "ODOMETER", "expression": "((B6*65536)+(B7*256)+B9)", "unit": "km", "class": "distance", "period": 5000 }
          ]
        },
        {
          "pid": "22 1E 32",
          "pid_init": "ATSH7E5;",
          "parameters": [
            {
              "name": "ENERGY_CHARGE_kWh",
              "expression": "((B12*16777216)+(B13*65536)+(B14*256)+B15)/8583.07212",
              "unit": "kWh",
              "class": "energy",
              "period": 8000
            }
          ]
        },
        {
          "pid": "22 22 E0",
          "pid_init": "ATSH714;ATCRA77E;",
          "parameters": [
            { "name": "RANGE_KM", "expression": "(B4*256)+B5", "unit": "km", "class": "range", "min": "0", "max": "350", "period": 4000 }
          ]
        }
      ]
    }
  ]
}
```


Download file: [vehicle_profiles.json](https://github.com/user-attachments/files/21836607/vehicle_profiles.json)
Test it with: `http://<wican-ip>/autopid_data`
